docs/user/sbt.rst

```diff
@@ -418,6 +418,81 @@ transparent to the library user.
 This feature can be used in combination with the feature above that
 allows native code in your application.
 
+EXPERIMENTAL: Conditional compilation of native glue code
+---------------------------------------------------------
+
+This is an **experimental** feature. Scala Native may deprecate or
+remove this feature with little or no warning.
+
+Libraries developed with "glue" code as described in the above section
+can cause compilation errors when all the following conditions occur:
+
+1. The library and/or header files are not installed
+
+2. The dependency is in the library users' build
+
+3. The code that uses the "glue" code is not called by the application
+   or library
+
+If the glue "code" is being called, then the library and headers need to
+be installed to compile your application otherwise errors are expected.
+
+Scala Native code can include the annotation ``@link("z")`` for example
+that says link with the ``z`` library. The compiler will add a link
+option for this library to the linking phase of the build if the code
+with the annotation is used. See :ref:`interop`,
+`Linking with native libraries` section for more information.
+
+This **experimental** feature has been added so the users of your published
+library can avoid the error described above. Use the following procedure to
+implement this feature.
+
+1. Add a Scala Native deployment descriptor to your library. For the
+purposes of this example assume the library is ``z``. The properties file
+must be named ``scala-native.properties`` and must be put in the base of the
+``src/main/resources/scala-native`` directory.
+
+2. Add the following content to your new ``scala-native.properties`` file.
+Note that if your library has more that one library you can add a comma
+delimited list of libraries. If desired, the comments are not needed.
+
+.. code-block:: properties
+
+    # configuration for glue code
+    # defines SCALANATIVE_LINK_Z if @link("z") annnotation is used (found in NIR)
+    # libraries used, comma delimited
+    nir.link.names = z
+
+3. Now in your native "glue" code add the following. The macro is named
+``SCALANATIVE_LINK_`` plus the uppercased name of the library.
+
+.. code-block:: c
+
+    #ifdef SCALANATIVE_LINK_Z
```

Comment:
I think we should decouple this feature from the `@link` annotation. As I mentioned, it's a really smart idea and I believe it can be useful in situations not involving linking to libraries.

For example, consider some bindings to a system/OS API that require some glue code. Maybe these APIs are only available in newer systems or otherwise optional. Or perhaps it's a non-trivial amount of glue code. io_uring is a great example of this: it's only available in newer Linux kernels and requires substantial glue (see https://github.com/armanbilge/fs2-io_uring/issues/62).

This is a case where (1) conditional compilation based on linktime reachability is valuable but (2) we are actually not linking to a particular library with an `-l` flag.

So perhaps it would be interesting to introduce a new annotation e.g. `@define("USER_CHOOSES_MACRO_NAME")`. If the target of that annotation is linktime reachable, then the macro is defined.

(I'm having a deja vu that maybe a similar idea like this was previously discussed 🤔 )

Reply:
You bring up some good points but it is hard to know all the ins and outs without a trial. This does put a small burden on the library developer but could really help the users as you adamantly pointed out in the original issue.

1. It should fix most of the cases you brought up.
2. It democratizes the feature so the internal build can use it plus users
3. It allow us to move the optional `z` lib to javalib where it belongs, See number 2
4. It simplifies the build by removing a bunch of filtering code
5. Downside: the optional `z` code does not get compiled / executed unless a scripted or other test checks it out. I hate to add one but maybe we need one?

>I think we should decouple this feature from the @link annotation. As I mentioned, it's a really smart idea and I believe it can be useful in situations not involving linking to libraries.

Well this is how this works now because if the link is in the NIR we can add the `-D` so the code compiles and the `-lxx` gets added to the link phase. This is actually a pretty cool side effect or feature of dead code elimination with NIR optimization.

I am not sure if you can just omit the `@link` but if you can then this feature would not help or affect the code at all. We can't look at the link flags because the users can put anything there.

>For example, consider some bindings to a system/OS API that requires some glue code. Maybe these APIs are only available in newer systems or otherwise optional. Or perhaps it's a non-trivial amount of glue code. io_uring is a great example of this: it's only available in newer Linux kernels and requires substantial glue (see https://github.com/armanbilge/fs2-io_uring/issues/62).

I think there are kernel versions and such that you can use to conditionally compile. Maybe?
https://stackoverflow.com/questions/56485820/conditional-compilation-based-on-functionality-in-linux-kernel-headers

I am not sure what other features could help in this area. I would be glad to look at specific issue with you and see if we could do more. The tricky thing is deciding how to give flexibility without too much complexity and to allow the simple use cases too.

I could think of a general way to handle the GC code or I would have done that too :-)

Reply:
Yes, I am saying we should make it possible to add the `-D` without the `-l`. There's no reason that adding macro definitions needs to be coupled to library linking. It's an independent and useful feature in its own right, which is exactly why I like it so much.

Reply:
Sure, we could add all sorts of things to the descriptor I think without adding too much complexity.

Reply:
I was thinking about this and I think you could leave out the `@link` annotation in your code and add the conditional compilation. Then have the users add the `-DSCALANATIVE_LINK_<LIB>` command and add the `-l<lib>` option if decide to use the feature. I haven't tried this but you could try in out even without this feature.